libsolidity/analysis/TypeChecker.cpp

```diff
@@ -183,9 +184,27 @@ void TypeChecker::checkContractDuplicateFunctions(ContractDefinition const& _con
 			msg
 		);
 	}
-	for (auto const& it: functions)
+
+	findDuplicateDefinitions(functions, "Function with same name and arguments defined twice.");
+}
+
+void TypeChecker::checkContractDuplicateEvents(ContractDefinition const& _contract)
+{
+	/// Checks that two events with the same name defined in this contract have different
+	/// argument types
+	map<string, vector<EventDefinition const*>> events;
+	for (EventDefinition const* event: _contract.events())
+		events[event->name()].push_back(event);
+
+	findDuplicateDefinitions(events, "Event with same name and arguments defined twice.");
+}
+
+template <class T>
+void TypeChecker::findDuplicateDefinitions(map<string, vector<T>> const& _definitions, string _message)
```

Comment:
Hm, maybe it could also work by using `CallableDeclaration const*` (the parent of both `EventDefinition` and `FunctionDefinition`) to avoid a template.

Reply:
The problem with that approach is that the `FunctionType` constructor called a couple of lines below is only defined for `EventDefinition` and `FunctionDefinition`, not for the generic `CallableDeclaration`.

I am far from a C++ expert, so if there is a simple way to fix that, I am all ears.

Reply:
True. It would be possible add another helper in `FunctionType` to accept `CallableDeclaration`, but perhaps the template is more cleaner for now.